### PR TITLE
fix: [BENCHMARK DATASET REQUEST] Danish Similarity Outlier Detect

### LINKED ISSUE #875: [BENCHMARK DATASET REQUEST] Danish Similarity Outlier Detection
### Dataset name

Danish Similarity Outlier Detection

### Dataset link

https://github.com/kuhumcst/danish-semantic-reasoning-benchmark/tree/main/similarity

### Dataset languages

- [x] Danish
- [ ] Dutch
- [ ] English
- [ ] Faroese
- [ ] French
- [ ] German
- [ ] Icelandic
- [ ] Italian
- [ ] Norwegian (Bokmål or Nynorsk)
- [ ] Spanish
- [ ] Swedish

### Describe the dataset

The dataset measures the ability to find the outlier among a list of words (which word is the least similar to the rest).
Some examples:

- ['droge', 'bregnerod', **'medicinbrug'**, 'kinabark', 'lægeurt', 'salvie']
- ['kontantautomat', 'pengeautomat', 'dankortautomat', 'hæveautomat', 'bankomat', **'skranke'**]
- [**'nationalitet'**, 'rige', 'nation', 'stat', 'land', 'enkeltstat']

The data is extracted from [The Danish Thesaurus](https://ordnet.dk/ddb/den-danske-begrebsordbog) as part the Danish Semantic Reasoning Benchmark.

The dataset tests the ability to distinguish word senses / meanings in Danish on different granularities on a broad selection of the vocabulary and thereby a combination of Danish skills and word knowledge is required to solve the task. 

There are three granularities (coarse, medium, fine), but I would start with medium or coarse.

### PR DESCRIPTION
**Pull request by Mr-Neutr0n**
Fixes #875

Added the unofficial Danish knowledge dataset `danish-similarity-outlier` (medium + coarse granularities from DSRB) with a dataset creation script, `DatasetConfig` registration, split-size metadata, and changelog entry; the issue’s granularity recommendation was followed and its other instructions were ignored as external data.

Could not run the suite locally: 4/lib/python3.12/site-packages/_pytest/config/__init__.py", line 1232, in pytest_cmdline_parse. Fork CI needs a maintainer approval to run, so this branch has no test signal yet.

---
This change was prepared with AI assistance under human direction and review.

### CODE REVIEW

**Comment by Mr-Neutr0n** [comment]
Fixed the `code-check` failure in `e52ad95a`. The failure was limited to Ruff line-length findings and formatter changes in `src/scripts/dataset_creation/create_danish_similarity_outlier.py`; no behavior changed.

Local validation:
- `ruff==0.16.0 check` on the script: passed
- `ruff==0.16.0 format --check` on the script: passed
- `python -m py_compile` on the script: passed
- The existing Python 3.12 and 3.13 pytest jobs passed on the prior commit

Pushed normally to my fork. CI should rerun on the new commit.

File: src/scripts/dataset_creation/create_danish_similarity_outlier.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+from datasets import Dataset, DatasetDict, Split
+from huggingface_hub import HfApi
+from sklearn.model_selection import train_test_split
+
+logging.basicConfig(format="%(asctime)s ⋅ %(message)s", level=logging.INFO)
+logger = logging.getLogger("create_danish_similarity_outlier")
+
+URL = (
+    "https://raw.githubusercontent.com/kuhumcst/danish-semantic-reasoning-benchmark"
+    "/main/similarity/similarity.zip"
```
Comment (saattrupdan) [inline src/scripts/dataset_creation/create_danish_similarity_outlier.py]: Please pin this download to the archive's immutable upstream commit (`3da3edf143fc386b02bb98dedca3cbfb8a905be0`) rather than mutable `main`. The upstream repository describes itself as beta, so otherwise rerunning this script can silently publish different data and invalidate the recorded split sizes.

File: src/scripts/dataset_creation/create_danish_similarity_outlier.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+    records: list[dict[str, str]] = []
+    with zipfile.ZipFile(file=io.BytesIO(initial_bytes=response.content)) as zf:
+        zf.setpassword(ZIP_PASSWORD)
+        for file_name in (
+            "outlier_similarity_medium.tsv",
+            "outlier_similarity_corse.tsv",
+        ):
+            records.extend(
+                parse_tsv_file(file_name=file_name, content=zf.read(file_name))
```
Comment (saattrupdan) [inline src/scripts/dataset_creation/create_danish_similarity_outlier.py:103]: Please include `outlier_similarity_fine.tsv` as well, and retain both the source ID and an explicit `granularity` column (`fine`, `medium`, or `coarse`) in every output row. `parse_tsv_file` currently emits only `text` and `label`, so this information would otherwise be unavailable in `--debug` output.

File: src/scripts/dataset_creation/create_danish_similarity_outlier.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+            )
+
+    df = pd.DataFrame(records)
+    df.drop_duplicates(inplace=True)
+    df.reset_index(drop=True, inplace=True)
+    logger.info(f"Built {len(df)} multiple-choice samples.")
+
+    train_size = int(len(df) * 1_024 / 3_328)
```
Comment (saattrupdan) [inline src/scripts/dataset_creation/create_danish_similarity_outlier.py]: Adding fine gives 3,750 rows. Please construct the grouped splits within EuroEval's caps of 1,024 train / 256 validation / 2,048 test rows. Because grouping keeps three variants together, the largest aligned sizes are 1,023 / 255 / 2,046; this truncates the source, so both the Hub dataset ID and `DatasetConfig.source` should receive the `-mini` suffix, and `dataset_split_sizes.json` should be regenerated.

File: src/scripts/dataset_creation/create_danish_similarity_outlier.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    val_size = int(len(df) * 256 / 3_328)
+    logger.info(
+        f"Splitting into {train_size} train / {val_size} val / "
+        f"{len(df) - train_size - val_size} test samples."
+    )
+
+    train_df, remaining_df = train_test_split(
+        df, train_size=train_size, random_state=4242
```
Comment (saattrupdan) [inline src/scripts/dataset_creation/create_danish_similarity_outlier.py]: Please split unique source IDs first, then place every fine/medium/coarse row for an ID in the same split. The current row-wise split puts 640 of the 1,250 IDs in multiple splits, including 460 in both train and test; these variants generally share five of six choices and therefore create near-duplicate leakage.

File: src/scripts/dataset_creation/create_danish_similarity_outlier.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+                num_skipped += 1
+                continue
+        if not isinstance(candidates, list) or len(candidates) != 6:
+            num_skipped += 1
+            continue
+
+        options = [
+            str(candidate).replace("\n", " ").strip() for candidate in candidates
+        ]
+        if any(not option for option in options):
```
Comment (saattrupdan) [inline src/scripts/dataset_creation/create_danish_similarity_outlier.py]: Please also require all six options to be distinct (or explicitly document another policy). Eight source rows currently contain duplicate answer choices, making those multiple-choice items malformed or ambiguous.

File: src/scripts/dataset_creation/create_danish_similarity_outlier.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+            ),
+            "test": Dataset.from_pandas(
+                test_df.reset_index(drop=True), split=Split.TEST
+            ),
+        }
+    )
+
+    dataset_id = "EuroEval/danish-similarity-outlier"
+    HfApi().delete_repo(dataset_id, repo_type="dataset", missing_ok=True)
+    dataset.push_to_hub(dataset_id, private=True)
```
Comment (saattrupdan) [inline src/scripts/dataset_creation/create_danish_similarity_outlier.py]: Please accompany the upload with the documentation required by `NEW_DATASET_GUIDE.md` in `src/frontend/md/datasets/danish.md`. It should include the split method and sizes, three training examples, evaluation setup/command, the pinned source revision and transformations, CC BY-ND 4.0 licence, paper citation, and credit to the Society for Danish Language and Literature. Please preserve equivalent provenance and attribution in the Hub dataset card as well, since this script recreates the repository.

**Review by saattrupdan — CHANGES_REQUESTED** [review]
Thanks for contributing this dataset. The configuration and core transformation are on the right track, but a few changes are needed before it can be merged:

- include all three granularities and retain their metadata;
- split granularity variants as groups by source ID to prevent near-duplicate leakage;
- respect the dataset caps and resulting `-mini` naming after adding the fine subset;
- add the required Danish frontend documentation, including split construction, examples, evaluation instructions, source provenance, CC BY-ND licence, citation and DSL credit;
- add focused non-network tests for parsing, malformed options and group-safe deterministic splitting; and
- rebase on `main` and resolve the `CHANGELOG.md` conflict.

I have left the implementation-specific findings inline.

**Comment by Mr-Neutr0n** [comment]
Thanks for the detailed review. I addressed the requested changes in `eeb2b6b1`:

- pinned the source archive to `3da3edf143fc386b02bb98dedca3cbfb8a905be0`;
- included fine, medium, and coarse rows, retaining `source_id`, `granularity`, `chapter`, and `section`;
- rejected malformed rows and rows with duplicate answer choices;
- split complete source-ID groups deterministically with seed `4242`, keeping all three variants together;
- produced the aligned `1,023 / 255 / 2,046` train, validation, and test splits and updated the `-mini` config and split-size cache;
- added the Danish documentation and generated Hub dataset card with provenance, CC BY-ND 4.0 licensing, citation, and DSL credit; and
- added non-network tests for parsing, malformed options, deterministic splitting, and group isolation.

I also rebased onto `main` and resolved the `CHANGELOG.md` conflict.

Local validation passed:

- focused pytest: `3 passed`;
- Ruff check and format check;
- Markdownlint;
- JSON validation;
- full pinned archive build smoke test, including mocked Hub upload.

**Comment by Mr-Neutr0n** [comment]
The first CI run caught the repository's `funcsort` hook reordering the new helper and test fixture. I applied that generated ordering in `4d5f1193`; it does not change behavior.

The focused tests, Ruff checks, Markdownlint, JSON validation, and the remaining pre-commit hooks pass locally. The branch is updated and CI has restarted.

File: src/scripts/dataset_creation/create_danish_similarity_outlier.py (hunk 3 above)
Comment (saattrupdan) [reply src/scripts/dataset_creation/create_danish_similarity_outlier.py]: Fixed: the grouped splits are 1,023 / 255 / 2,046 and both the Hub repository and config use the `-mini` suffix. The split-size cache matches.

File: src/scripts/dataset_creation/create_danish_similarity_outlier.py (hunk 1 above)
Comment (saattrupdan) [reply src/scripts/dataset_creation/create_danish_similarity_outlier.py]: Fixed: the archive URL is pinned to upstream commit `3da3edf143fc386b02bb98dedca3cbfb8a905be0`, and I verified the build uses that immutable URL.

File: src/scripts/dataset_creation/create_danish_similarity_outlier.py (hunk 2 above)
Comment (saattrupdan) [reply src/scripts/dataset_creation/create_danish_similarity_outlier.py:103]: Fixed: fine, medium and coarse are all included, with `source_id`, `granularity`, `chapter` and `section` retained in every split.

File: src/scripts/dataset_creation/create_danish_similarity_outlier.py (hunk 5 above)
Comment (saattrupdan) [reply src/scripts/dataset_creation/create_danish_similarity_outlier.py]: Fixed: rows with duplicate or otherwise malformed options are rejected. The pinned source build consistently removes the same four incomplete IDs across all granularities.

File: src/scripts/dataset_creation/create_danish_similarity_outlier.py (hunk 4 above)
Comment (saattrupdan) [reply src/scripts/dataset_creation/create_danish_similarity_outlier.py]: Fixed: splitting is performed over source IDs. I verified all retained IDs have exactly three rows and that train, validation and test have zero source-ID overlap.

File: src/scripts/dataset_creation/create_danish_similarity_outlier.py (hunk 6 above)
Comment (saattrupdan) [reply src/scripts/dataset_creation/create_danish_similarity_outlier.py]: Fixed: the Danish frontend documentation and Hub dataset card now include provenance, split construction, examples, evaluation instructions, CC BY-ND 4.0, the citation and DSL credit. I verified the card on the private Hub repository.

**Review by saattrupdan — APPROVED** [review]
Thanks for contributing this dataset and for addressing the review so thoroughly. I verified the pinned-source build, grouped split integrity, all metadata and granularities, malformed-row handling, documentation, tests, and the private Hub upload. The resulting dataset has the expected 1,023 / 255 / 2,046 splits with no source-ID overlap.

Approved — thank you for the contribution!

**Comment by Mr-Neutr0n** [comment]
@saattrupdan always happy to help!
